### PR TITLE
prefilling start forms

### PR DESCRIPTION
**Pull request by luis-soares-ritense**

### CODE REVIEW

File: form-flow-valtimo/src/main/kotlin/com/ritense/valtimo/formflow/FormFlowProcessLinkActivityHandler.kt [hunk 1]
```diff
@@ -25,6 +25,7 @@ import com.ritense.processlink.web.rest.dto.ProcessLinkActivityResult
 import com.ritense.valtimo.formflow.domain.FormFlowProcessLink
 import org.camunda.bpm.engine.RuntimeService
 import org.camunda.bpm.engine.task.Task
+import java.util.*
```
Comment (theo-ritense) [inline form-flow-valtimo/src/main/kotlin/com/ritense/valtimo/formflow/FormFlowProcessLinkActivityHandler.kt:28]: Please avoid wildcards in production code.

File: form-link/src/main/java/com/ritense/formlink/service/impl/CamundaFormAssociationService.java [hunk 2]
```diff
@@ -90,6 +91,18 @@ public CamundaFormAssociationService(
         this.formFieldDataResolvers = formFieldDataResolvers;
     }
 
+    private static <T> Collector<T, ?, Optional<T>> singleElementCollector() {
```
Comment (theo-ritense) [inline form-link/src/main/java/com/ritense/formlink/service/impl/CamundaFormAssociationService.java:94]: Can't this whole collector be replaced by using `.findFirst()` on the stream?

File: form-link/src/main/java/com/ritense/formlink/web/rest/FormAssociationResource.java [hunk 3]
```diff
@@ -21,6 +21,7 @@
 import java.util.Optional;
 import java.util.UUID;
 import org.springframework.http.ResponseEntity;
+import org.springframework.web.bind.annotation.RequestParam;
```
Comment (theo-ritense) [inline form-link/src/main/java/com/ritense/formlink/web/rest/FormAssociationResource.java:24]: I don't see this new import being used.

File: form-link/src/main/java/com/ritense/formlink/web/rest/impl/CamundaFormAssociationResource.java [hunk 4]
```diff
@@ -72,9 +72,10 @@ public ResponseEntity<JsonNode> getPreFilledFormDefinition(
     @Override
     @GetMapping(value = "/v1/form-association/form-definition", params = {"processDefinitionKey"})
     public ResponseEntity<JsonNode> getStartEventFormDefinitionByProcessDefinitionKey(
-        @RequestParam String processDefinitionKey
+        @RequestParam String processDefinitionKey,
+        @RequestParam(required = false) Optional<UUID> documentId
     ) {
-        return formAssociationService.getStartEventFormDefinition(processDefinitionKey)
+        return formAssociationService.getStartEventFormDefinition(processDefinitionKey,documentId)
```
Comment (theo-ritense) [inline form-link/src/main/java/com/ritense/formlink/web/rest/impl/CamundaFormAssociationResource.java:78]:
```suggestion
        return formAssociationService.getStartEventFormDefinition(processDefinitionKey, documentId)
```

File: form-link/src/main/java/com/ritense/formlink/service/impl/CamundaFormAssociationService.java [hunk 5]
```diff
@@ -122,16 +135,17 @@ public Optional<JsonNode> getFormDefinitionByFormLinkId(
             processDefinitionKey,
             formLinkId
         ).flatMap(
-            camundaFormAssociation -> Optional.of(buildFormDefinition(camundaFormAssociation))
+            camundaFormAssociation -> Optional.of(buildFormDefinition(camundaFormAssociation,Optional.empty()))
```
Comment (theo-ritense) [inline form-link/src/main/java/com/ritense/formlink/service/impl/CamundaFormAssociationService.java:138]:
```suggestion
            camundaFormAssociation -> Optional.of(buildFormDefinition(camundaFormAssociation, Optional.empty()))
```

File: form-flow-valtimo/src/test/kotlin/com/ritense/valtimo/formflow/FormFlowProcessLinkActivityHandlerIntTest.kt [hunk 6]
```diff
@@ -133,7 +133,7 @@ internal class FormFlowProcessLinkActivityHandlerIntTest: BaseIntegrationTest()
             activityType = ActivityTypeWithEventName.START_EVENT_START,
             formFlowDefinitionId = formFlowDefinition?.id.toString())
 
-        val result = processLinkActivityHandler.getStartEventObject("",processLink)
+        val result = processLinkActivityHandler.getStartEventObject("",null,processLink)
```
Comment (theo-ritense) [inline form-flow-valtimo/src/test/kotlin/com/ritense/valtimo/formflow/FormFlowProcessLinkActivityHandlerIntTest.kt:136]:
```suggestion
        val result = processLinkActivityHandler.getStartEventObject("", null, processLink)
```

File: form-link/src/test/java/com/ritense/formlink/web/rest/CamundaFormAssociationResourceTest.java [hunk 7]
```diff
@@ -125,7 +125,7 @@ public void shouldReturn200WithStartFormByProcessDefinitionKey() throws Exceptio
         final ObjectNode jsonNode = JsonNodeFactory.instance.objectNode();
         when(camundaFormAssociationService
             .getStartEventFormDefinition(
-                eq(camundaProcessFormAssociation.getProcessDefinitionKey())
+                eq(camundaProcessFormAssociation.getProcessDefinitionKey()),eq(Optional.empty())
```
Comment (theo-ritense) [inline form-link/src/test/java/com/ritense/formlink/web/rest/CamundaFormAssociationResourceTest.java:128]:
```suggestion
                eq(camundaProcessFormAssociation.getProcessDefinitionKey()), eq(Optional.empty())
```

File: form/src/main/kotlin/com/ritense/form/processlink/FormProcessLinkActivityHandler.kt [hunk 8]
```diff
@@ -24,6 +24,7 @@ import com.ritense.processlink.domain.ProcessLink
 import com.ritense.processlink.service.ProcessLinkActivityHandler
 import com.ritense.processlink.web.rest.dto.ProcessLinkActivityResult
 import org.camunda.bpm.engine.task.Task
+import java.util.*
```
Comment (theo-ritense) [inline form/src/main/kotlin/com/ritense/form/processlink/FormProcessLinkActivityHandler.kt:27]: Please avoid wildcards in production code.

File: form/src/main/kotlin/com/ritense/form/processlink/FormProcessLinkActivityHandler.kt [hunk 9]
```diff
@@ -49,10 +50,11 @@ class FormProcessLinkActivityHandler(
 
     override fun getStartEventObject(
         processDefinitionId: String,
+        documentId: UUID?,
         processLink: ProcessLink
     ): ProcessLinkActivityResult<FormTaskOpenResultProperties> {
         processLink as FormProcessLink
-        val formDefinition = formDefinitionService.getFormDefinitionById(processLink.formDefinitionId).orElseThrow()
+        val formDefinition = prefillFormService.getPrefilledFormDefinition(processLink.formDefinitionId,documentId)
```
Comment (theo-ritense) [inline form/src/main/kotlin/com/ritense/form/processlink/FormProcessLinkActivityHandler.kt:57]:
```suggestion
        val formDefinition = prefillFormService.getPrefilledFormDefinition(processLink.formDefinitionId, documentId)
```

File: process-link/src/main/kotlin/com/ritense/processlink/service/ProcessLinkActivityService.kt [hunk 10]
```diff
@@ -43,13 +43,16 @@ open class ProcessLinkActivityService(
...
+        processDefinitionId: String,
+        documentId: UUID?
+        ): ProcessLinkActivityResult<*>? {
         val processLink = processLinkService.getProcessLinksByProcessDefinitionIdAndActivityType(processDefinitionId,
             ActivityTypeWithEventName.START_EVENT_START)
         var result: ProcessLinkActivityResult<*>? = null
         processLinkActivityHandlers.forEach {
             if(it.supports(processLink)){
-                result = it.getStartEventObject(processDefinitionId, processLink)
+                result = it.getStartEventObject(processDefinitionId,documentId, processLink)
```
Comment (theo-ritense) [inline process-link/src/main/kotlin/com/ritense/processlink/service/ProcessLinkActivityService.kt:55]:
```suggestion
                result = it.getStartEventObject(processDefinitionId, documentId, processLink)
```

File: process-link/src/main/kotlin/com/ritense/processlink/web/rest/ProcessLinkTaskResource.kt [hunk 11]
```diff
@@ -41,7 +42,10 @@ class ProcessLinkTaskResource(
         }}
 
     @GetMapping(value = ["/v1/process-definition/{processDefinitionId}/start-form"])
-    fun getFormDefinition(@PathVariable processDefinitionId: String): ResponseEntity<Any> {
-        return ResponseEntity.ok(processLinkActivityService.getStartEventObject(processDefinitionId))
+    fun getFormDefinition(
+        @PathVariable processDefinitionId: String,
+        @RequestParam(required = false) documentId: UUID?
+    ): ResponseEntity<Any> {
+        return ResponseEntity.ok(processLinkActivityService.getStartEventObject(processDefinitionId,documentId))
```
Comment (theo-ritense) [inline process-link/src/main/kotlin/com/ritense/processlink/web/rest/ProcessLinkTaskResource.kt:49]:
```suggestion
        return ResponseEntity.ok(processLinkActivityService.getStartEventObject(processDefinitionId, documentId))
```

File: form-link/src/main/java/com/ritense/formlink/service/impl/CamundaFormAssociationService.java (hunk 2 above)
Comment (luis-soares-ritense) [reply form-link/src/main/java/com/ritense/formlink/service/impl/CamundaFormAssociationService.java:94]: don't know what code is this but i didn't write it. for some reason it appears as i did it. but i didn't.

**Review by Klaas-Ritense — APPROVED** [review]

**Pull request merged**